keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,146 @@
+
+Consider a cluster that has nodes in multiple regions. We want to ensure that a nginx deployment has 2 replicas in one region and 1 replica in the other one. Furthermore they should support both HTTP as well as HTTPS, hence need a list of port definitions. These parameters are described by `list` and `dict` types:
+
+```yaml
+parameters:
+  - name: ports
+    description: Ports
+    type: list
+    default: [ 80, 443 ]
+  - name: topology
+    description: Node topology
+    type: dict
+    default: |
+      - region: us-east-1:
+        replicas: 2
+      - region: us-west-1:
+        replicas: 1
+```
+
+The deployment template then iterates over the items of the respective parameters:
+
+```yaml
+{{ $ports := .Params.ports }}
+{{ range .Params.topology }}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+  labels:
+    app.kubernetes.io/name: nginx
+spec:
+  replicas: {{ .replicas }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: nginx
+  template:
+    metadata:
+      name: nginx
+      labels:
+        app.kubernetes.io/name: nginx
+    spec:
+      nodeSelector:
+        topology.kubernetes.io/region: {{ .region }}
+      containers:
+      - name: nginx
+        image: nginx:stable
+        ports:
+        {{ range $ports }}
+        - containerPort: {{ . }}
+        {{ end }}
+{{ end }}
+```
+
+### Provide Parameter Validation
+
+Adding types for lists and dictionaries provides access to the items of these structures in templates. Adding types like integers won't provide benefits for templating, but would allow for input value validation. KUDO could, e.g., validate that an integer isn't too large, that a list doesn't have too many entries, or that a string isn't too long. Adding parameter validation is independent from adding parameter types.
+
+Adding validation rules to parameters could look like this:
+
+```yaml
+parameters:
+  - name: replicas
+    description: Number of replicas
+    type: integer
+    minimum: 1
+    maximum: 3
+  - name: instance-name
+    description: Name of the instance
+    regex: ^\w-_*$
+    max-length: 40
+```
+
+### Alternative Considered: Convert YAML in string to dictionaries
+
+A template function `fromYAML` could be used to convert a parameter value to a dictionary. The example above would then change to
+
+```yaml
+{{ $topology := (fromYAML .Params.topology) }}
+{{ range $topology }}
+...
+{{ end }}
+```
+
+In KUDO, this function would work the same way as the `dict` type conversion mentioned above. The difference is that the conversion has to be explicitly done by the operator developer in the template.
```

Comment:
That's not the only downside, is it? It would also be hard to perform validation, right?

Reply:
Yes, let me add this.